File: pascal/classes/passing_object_to_function.pas
program PassingObjectToFunction;

type Book = object
  public
    title: string;
    author: string;
    isbn: string;

    constructor create(book_title: string; book_author: string; book_isbn: string);
end;

type
  bkptr = ^Book;

constructor Book.create(book_title: string; book_author: string; book_isbn: string);
begin
  title := book_title;
  author := book_author;
  isbn := book_isbn;
end;

procedure describe_book_by_value(described_book: Book);
begin
  writeln('This is a book [by value]:');
  write('title: ');
  writeln(described_book.title);
  write('author: ');
  writeln(described_book.author);
  write('isbn: ');
  writeln(described_book.isbn);
  writeln();
end;

procedure describe_book_by_pointer(described_book: bkptr);
begin
  writeln('This is a book [by pointer]:');
  write('title: ');
  writeln(described_book^.title);
  write('author: ');
  writeln(described_book^.author);
  write('isbn: ');
  writeln(described_book^.isbn);
  writeln();
end;

var
  novel: Book;

begin
  novel.create('Decameron', 'Giovanni Boccacio', '978-83-7779-436-4');

  writeln('The book presentation:');
  writeln();
  describe_book_by_value(novel);
  describe_book_by_pointer(@novel);
end.

File: c++/classes/passing_object_to_function.cpp
#include <iostream>

class Book {
public:
  const char *title;
  const char *author;
  const char *isbn;

  Book(const char *book_title, const char *book_author, const char *book_isbn) {
    title = book_title;
    author = book_author;
    isbn = book_isbn;
  }
};

void describe_book_by_value(Book book) {
  std::cout << "This is a book [by value]:\n";
  std::cout << "title: " << book.title << std::endl;
  std::cout << "author: " << book.author << std::endl;
  std::cout << "ISBN: " << book.isbn << std::endl << std::endl;
}

void describe_book_by_pointer(Book *book) {
  std::cout << "This is a book [by pointer]:\n";
  std::cout << "title: " << book->title << std::endl;
  std::cout << "author: " << book->author << std::endl;
  std::cout << "ISBN: " << book->isbn << std::endl << std::endl;
}

void describe_book_by_reference(Book &book) {
  std::cout << "This is a book [by reference]:\n";
  std::cout << "title: " << book.title << std::endl;
  std::cout << "author: " << book.author << std::endl;
  std::cout << "ISBN: " << book.isbn << std::endl << std::endl;
}

int main() {
  Book novel("Decameron", "Giovanni Boccacio", "978-83-7779-436-4");

  std::cout << "The book presentation:\n\n";
  describe_book_by_value(novel);
  describe_book_by_pointer(&novel);
  describe_book_by_reference(novel);
}
